src/main/java/com/team766/controllers/ExtendableCanSparkMaxSmartMotionPIDController.java

```diff
@@ -0,0 +1,283 @@
+package com.team766.controllers;
+
+import com.team766.config.ConfigFileReader;
+import com.team766.hal.RobotProvider;
+import com.team766.library.SetValueProvider;
+import com.team766.library.SettableValueProvider;
+import com.team766.library.ValueProvider;
+import com.team766.logging.Logger;
+import com.team766.logging.LoggerExceptionUtils;
+import edu.wpi.first.math.MathUtil;
+import com.revrobotics.SparkMaxAbsoluteEncoder;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.ControlType;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.SparkMaxAbsoluteEncoder.Type;
+import java.io.IOError;
+import javax.swing.text.DefaultStyledDocument.ElementSpec;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.revrobotics.CANSparkMax;
+import com.team766.framework.Mechanism;
+import com.team766.hal.MotorController;
+import com.team766.library.RateLimiter;
+import com.team766.controllers.pidstate.*;
+import com.team766.framework.Exceptions.*;
+
+
+public abstract class ExtendableCanSparkMaxSmartMotionPIDController {
+		/*
+		 * These are the hardware variables for the hardware we have on the robot.
+		 * It includes a CanSparkMax, MotorController of any type, and a SparkMaxAbsoluteEncoder.
+		 * @author Max Spier - 9/9/2023
+		 */
+		private MotorController mc;
+		private CANSparkMax csm;
+		private SparkMaxAbsoluteEncoder abs;
+	
+		/*
+		 * These are PID related variables that we will store locally to run PIDs
+		 * We also create an object for the PIDController
+		 * @author Max Spier - 9/9/2023
+		 */
+		private SparkMaxPIDController pid;
+		private static double deadzone = 0;
+		private static double setPointPosition = 0;
+		private static double comboOfTimesInsideDeadzone = 0;
+		private static double minPos = 0;
+		private static double maxPos = 0;
+	
+		/*
+		 * These are some variables that we can use so that we can reset the motor controller
+		 * We use this so that we can get an accurate ratio for the hall encoder units to absolute encoder units.
+		 * @author Max Spier - 9/9/2023
+		 */
+		private double encoderUnitsPerOneAbsolute = 0;
+
+		private double curAntigrav = 0;
+	
+	
+		/*
+		* The state of the PID controller so that we know whether to call the respective method for each state.
+		* For more documentation, see the PIDSTATE class.
+		* @author Max Spier - 9/9/2023
+		*/
+		
+		private PIDSTATE theState = PIDSTATE.OFF;
+
+	/*
+	 * Default constructor for the class. This is used to create a new CanSparkMaxSmartMotionPIDController object
+	 * @param configName the configuration name for the motorcontroller
+	 * @param absEncoderOffset the offset (ranging from 0.000 - 0.999) of the absolute encoder
+	 * @param absEncoderOffsetForZeroEncoderUnits the offset that the absolute encoder should have in order for the relative encoder on the motorcontroller to equal zero.
```

Comment:
in what case would `absEncoderOffsetForZeroEncoderUnits` not be 0?

Reply:
We want the arm to have 0 hall encoder units when the absolute encoder is at 0.362, so 0.362 is the ```absEncoderOffsetForZeroEncoderUnits``` value.

Reply:
why wouldn't we set `absEncoderOffset` so that 0 is 0?

Reply:
We could if we wanted, but this makes it easier if, for example, a person installs the encoder wrong and it is off by 0.203 units, so we can use this to reset that. Even if we wanted 0 to be 0, it isn't much more work to add that in, and later we could also overload the constructor to make that default 0.

Reply:
the motor encoder does not have a persistent offset - it just measures changes in position. you call `mc.setSensorPosition(relEncoder);` in the constructor to set it to whatever starting value you want, and then the value changes from there as the sensor detects changes in position.

so the only way to install it wrong would be to have it inverted (so moving the mechanism results in increasing values instead of decreasing and vice versa).

Reply:
Maybe I don't understand what your saying -- lets talk on Saturday